### PR TITLE
Continue implementing incident maintenance feature under beta

### PR DESCRIPTION
**Pull request by hellcp-work**
We reuse the incident maintenance workflow feature when we have the beta flag active.
Continutation of #14629

### CODE REVIEW

File: src/api/spec/features/beta/webui/maintenance_workflow_spec.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
+require 'browser_helper'
+
+RSpec.describe 'MaintenanceWorkflow', js: true, vcr: true do
```
Comment (eduardoj) [inline src/api/spec/features/beta/webui/maintenance_workflow_spec.rb:3]: Setting `vcr: false` and running the test I obtain `Failed to change state: The source of package home:tom:branches:Update/cacti.openSUSE_11.4_Update is broken!` errors:
```
frontend@:/obs/src/api> bin/rspec spec/features/beta/webui/maintenance_workflow_spec.rb
To use retry middleware with Faraday v2.0+, install `faraday-retry` gem
                                                                                               
Randomized with seed 33506                                                                                                                                                                    

  1) MaintenanceWorkflow maintenance request without patchinfo when accepting request succeeds
     Failure/Error: expect(page).to have_text("Request #{bs_request_without_patchinfo.number} accepted")
       expected to find text "Request 1 accepted" in "Watchlist\nNotifications\nPlaces\nTasks 1\nYour Home Project\nAll Projects\nStatus Monitor\nCollapse sidebar\nFailed to change state: The source of package home:tom:branches:Update/cacti.openSUSE_11.4_Update is broken!\nRequest 1\nAs part of the beta program, we are currently working on a redesign of this page to help you in your daily work. Certain features might still be missing, but they will be implemented in the future. Give us feedback on this redesign at GitHub by opening an issue or talk to us on IRC in the #opensuse-buildservice on Libera.Chat - The OBS team\nRequest 1 new\nWatch\nCreated by Tamera (user_10) less than a minute ago\nThis is a Maintenance Incident\nPrevious\nIncident cacti.openSUSE_11.4_Update\nNext\nShowing #1 (of 2)\nAction details Seen\nSubmit update from package home:tom:branches:Update / cacti.openSUSE_11.4_Update to project MaintenanceProject\nRelease in project openSUSE:11.4:Update\nConversation\nBuild Results\nRPM Lint\nChanges\nMentioned Issues 0\nAdd Reviewer\nProject Maintainers\nmaintenance_coord\nDescription\nVoluptatem impedit veniam. Asperiores sed odit. Voluptates reiciendis adipisci.\nComments & History\nTamera (user_10) created this request less than a minute ago\nWrite\nPreview\nUSER\nYour Profile\nHome Project\nLogout\nLOCATIONS\nProjects\nSearch\nStatus Monitor\nHELP\nOpenBuildService.org\nDocumentation\nAPI Documentation\nCONTACT\nSupport\n@OBShq\nOPEN BUILD SERVICE IS SPONSORED BY\nThe Open Build Service is an openSUSE project."
     # ./spec/features/beta/webui/maintenance_workflow_spec.rb:55:in `block (4 levels) in <top (required)>'
 
  2) MaintenanceWorkflow maintenance request without patchinfo when accepting request creates maintenance incident project
     Failure/Error: expect(page).to have_text('MaintenanceProject:0')
       expected to find text "MaintenanceProject:0" in "Watchlist\nNotifications\nPlaces\nTasks 1\nYour Home Project\nAll Projects\nStatus Monitor\nCollapse sidebar\nFailed to change state: The source of package home:tom:branches:Update/cacti.openSUSE_11.4_Update is broken!\nRequest 1\nAs part of the beta program, we are currently working on a redesign of this page to help you in your daily work. Certain features might still be missing, but they will be implemented in the future. Give us feedback on this redesign at GitHub by opening an issue or talk to us on IRC in the #opensuse-buildservice on Libera.Chat - The OBS team\nRequest 1 new\nWatch\nCreated by Harriet (user_15) less than a minute ago\nThis is a Maintenance Incident\nPrevious\nIncident cacti.openSUSE_11.4_Update\nNext\nShowing #1 (of 2)\nAction details Seen\nSubmit update from package home:tom:branches:Update / cacti.openSUSE_11.4_Update to project MaintenanceProject\nRelease in project openSUSE:11.4:Update\nConversation\nBuild Results\nRPM Lint\nChanges\nMentioned Issues 0\nAdd Reviewer\nProject Maintainers\nmaintenance_coord\nDescription\nSunt accusamus et. Eligendi illum qui. Laudantium ex eius.\nComments & History\nHarriet (user_15) created this request less than a minute ago\nWrite\nPreview\nUSER\nYour Profile\nHome Project\nLogout\nLOCATIONS\nProjects\nSearch\nStatus Monitor\nHELP\nOpenBuildService.org\nDocumentation\nAPI Documentation\nCONTACT\nSupport\n@OBShq\nOPEN BUILD SERVICE IS SPONSORED BY\nThe Open Build Service is an openSUSE project."
     # ./spec/features/beta/webui/maintenance_workflow_spec.rb:59:in `block (4 levels) in <top (required)>'
                                               
 3/3 |================================================================================ 100 =================================================================================>| Time: 00:00:28 
                                               
Finished in 28.45 seconds (files took 5.55 seconds to load)
3 examples, 2 failures                                                                                                                                                                        
                                                                                                                                                                                              
Failed examples:
                                               
rspec ./spec/features/beta/webui/maintenance_workflow_spec.rb:58 # MaintenanceWorkflow maintenance request without patchinfo when accepting request creates maintenance incident project
rspec ./spec/features/beta/webui/maintenance_workflow_spec.rb:54 # MaintenanceWorkflow maintenance request without patchinfo when accepting request succeeds

Randomized with seed 35472
Coverage report generated for RSpec to /obs/src/api/coverage. 4024 / 32934 LOC (12.22%) covered.
Stopped processing SimpleCov as a previous error not related to SimpleCov has been detected
frontend@:/obs/src/api>                                                             
```

I tried again resetting the test database, with `RAILS_ENV=test bundle exec rake db:seed`. Same error.

Did I miss something else to set up?

**Review by eduardoj — DISMISSED** [review]
See previous comment.

File: src/api/spec/features/beta/webui/maintenance_workflow_spec.rb (hunk 1 above)
Comment (eduardoj) [reply src/api/spec/features/beta/webui/maintenance_workflow_spec.rb:3]: > Did I miss something else to set up?

Answering myself: yes. I didn't include the `GLOBAL_WRITE_THROUGH` variable. The tests pass with: `GLOBAL_WRITE_THROUGH=on bin/rspec spec/features/beta/webui/maintenance_workflow_spec.rb`.

File: src/api/spec/features/beta/webui/maintenance_workflow_spec.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  let!(:another_branched_package) { create(:package_with_file, project: branched_project, name: 'cacti.openSUSE_11.5_Update', file_name: 'DUMMY_FILE', file_content: 'boo#12345') }
+  let(:update_project) { create(:update_project, maintained_project: project, name: "#{project}:Update") }
+  let(:another_update_project) { create(:update_project, maintained_project: another_project, name: "#{another_project}:Update") }
+  let(:bs_request_without_patchinfo) do
+    create(:bs_request_with_maintenance_incident_actions, source_project_name: branched_project.name, source_package_names: [package.name], target_project_name: maintenance_project.name,
+                                                          target_releaseproject_names: [update_project.name, another_update_project.name])
+  end
+
+  before do
+    User.session = admin_user
```
[the remark this answers is not in the document]
Comment (hennevogel) [reply src/api/spec/features/beta/webui/maintenance_workflow_spec.rb:29]: And I think this should go down to the context

File: src/api/spec/features/beta/webui/maintenance_workflow_spec.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  let(:package) { create(:package_with_file, project: project, name: 'cacti') }
+  let!(:branched_package) { create(:package_with_file, project: branched_project, name: 'cacti.openSUSE_11.4_Update', file_name: 'DUMMY_FILE', file_content: 'boo#12345') }
+  let(:another_package) { create(:package_with_file, project: another_project, name: 'cacti') }
+  let!(:another_branched_package) { create(:package_with_file, project: branched_project, name: 'cacti.openSUSE_11.5_Update', file_name: 'DUMMY_FILE', file_content: 'boo#12345') }
+  let(:update_project) { create(:update_project, maintained_project: project, name: "#{project}:Update") }
+  let(:another_update_project) { create(:update_project, maintained_project: another_project, name: "#{another_project}:Update") }
+  let(:bs_request_without_patchinfo) do
+    create(:bs_request_with_maintenance_incident_actions, source_project_name: branched_project.name, source_package_names: [package.name], target_project_name: maintenance_project.name,
+                                                          target_releaseproject_names: [update_project.name, another_update_project.name])
+  end
```
Comment (hennevogel) [inline src/api/spec/features/beta/webui/maintenance_workflow_spec.rb:26]: I think you can simplify the setup of `update_project` and `another_update_project` with a block of code too. Like

```ruby
let(:another_update_project) do
  create(:project_with_repository, name: 'openSUSE:11.5')
  create(:package_with_file, project: another_project, name: 'cacti')
  create(:update_project, maintained_project: another_project, name: "#{another_project}:Update")
end
```

File: src/api/spec/features/beta/webui/maintenance_workflow_spec.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+        click_button('Accept request')
+      end
+
+      it 'succeeds' do
+        expect(page).to have_text("Request #{bs_request_without_patchinfo.number} accepted")
+      end
+
+      it 'creates maintenance incident project' do
+        expect(page).to have_text('MaintenanceProject:0')
```
Comment (hennevogel) [inline src/api/spec/features/beta/webui/maintenance_workflow_spec.rb:57]: You can also expect *real* changes, not only what is displayed. For instance that the request changes the state, that the `:0` project is created etc.

**Review by hennevogel — CHANGES_REQUESTED** [review]

File: src/api/spec/features/beta/webui/maintenance_workflow_spec.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  let(:update_project) { create(:update_project, maintained_project: project, name: "#{project}:Update") }
+  let(:another_update_project) { create(:update_project, maintained_project: another_project, name: "#{another_project}:Update") }
+  let(:bs_request_without_patchinfo) do
+    create(:bs_request_with_maintenance_incident_actions, source_project_name: branched_project.name, source_package_names: [package.name], target_project_name: maintenance_project.name,
+                                                          target_releaseproject_names: [update_project.name, another_update_project.name])
+  end
+
+  before do
+    User.session = admin_user
+    create(:maintenance_project_attrib, project: maintenance_project)
```
Comment (krauselukas) [inline src/api/spec/features/beta/webui/maintenance_workflow_spec.rb:30]: @hellcp-work isn't this already done in the factory of the maintenance_project? https://github.com/openSUSE/open-build-service/blob/852c1c2d45a54550e25a9e4ab807ceda3d128232/src/api/spec/factories/project.rb#L140

**Review by hennevogel — APPROVED** [review]

**Pull request merged**